frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -1305,6 +1305,114 @@ const IR::Node* TypeInference::postorder(IR::Concat* expression) {
     return expression;
 }
 
+const IR::Node* TypeInference::postorder(IR::Key* key) {
+    LOG1("TypeInference: visiting " << dbp(key));
+    // compute the type and store it in typeMap
+    auto vec = new IR::Vector<IR::Type>();
+    for (auto ke : *key->keyElements) {
+        auto kt = typeMap->getType(ke->expression);
+        if (kt == nullptr) {
+            LOG2("Bailing out for " << dbp(ke));
+            return key;
+        }
+        vec->push_back(kt);
+    }
+    auto keyTuple = new IR::Type_Tuple(Util::SourceInfo(), vec);
+    LOG2("Setting key type to " << dbp(keyTuple));
+    setType(key, keyTuple);
+    // installing also for the original because we cannot tell which one will survive in the ir
+    // \TODO: figure out the rules of survivorship
+    LOG2("Setting key type to " << dbp(getOriginal()));
+    setType(getOriginal(), keyTuple);
+    return key;
+}
+
+/**
+    typecheck a table initializer entry list
+
+    The invariants are:
+    - table keys must have been type checked before entries
+*/
+const IR::Node* TypeInference::postorder(IR::EntriesList* el) {
+    LOG1("TypeInference: visiting: " << el);
+    if (done()) return el;
+    auto table = findContext<IR::P4Table>();
+    BUG_CHECK(table != nullptr, "%1% entries not within a table", el);
+    const IR::Key* key = table->getKey();
+    if (key == nullptr) {
+        ::error("Could not find key for table %1%", table);
+        return el;
+    }
+
+    // the key must have been typechecked already.
+    if (!typeMap->contains(key)) {
+        ::error("Table initializer entries must occur after key definitions (table %1%)", table);
+        return el;
+    }
+    return el;
+}
+/**
+    typecheck a table initializer entry
+
+    The invariants are:
+    - table keys and entry keys must have the same length
+    - entry key elements must be compile time constants
+    - actionRefs in entries must be in the action list
+
+    Moreover, the EntriesList visitor should have checked for the table
+    invariants, and thus, those checks are now bug checks.
+*/
+const IR::Node* TypeInference::postorder(IR::Entry* entry) {
+    LOG1("TypeInference: visiting: " << entry);
+    if (done()) return entry;
+    auto table = findContext<IR::P4Table>();
+    BUG_CHECK(table != nullptr, "%1% entries not within a table", entry);
```

Comment:
If I move the ::error calls in pre-order I guess, I can just silently handle these errors, since I already printed them, and if I want to continue through other errors, I should just return.